crates/matrix-sdk-ui/src/timeline/queue.rs

```diff
@@ -96,78 +101,104 @@ pub(super) async fn send_queued_messages(
+        timeline
+            .update_event_send_state(
+                &msg.txn_id,
+                EventSendState::SendingFailed {
+                    // FIXME: Probably not exactly right
+                    error: Arc::new(matrix_sdk::Error::InconsistentState),
+                },
+            )
+            .await;
+        return;
     }
+
+    send_task.start(room, msg);
 }
 
-async fn handle_task_ready(
-    result: SendMessageResult,
+async fn handle_send_result(
+    send_result: SendMessageResult,
     send_task: &mut SendMessageTask,
     queue: &mut VecDeque<LocalMessage>,
-    timeline_inner: &TimelineInner,
+    timeline: &TimelineInner,
 ) {
-    match result {
-        SendMessageResult::Success { room } => {
+    match send_result {
+        SendMessageResult::Success { event_id, txn_id } => {
+            timeline.update_event_send_state(&txn_id, EventSendState::Sent { event_id }).await;
+
+            // Event was successfully sent, move on to the next queued event.
             if let Some(msg) = queue.pop_front() {
-                send_task.start(room, timeline_inner.clone(), msg);
+                send_task.start(timeline.room().clone(), msg);
             }
         }
-        SendMessageResult::SendingFailed => {
-            // Timeline items are marked as failed / cancelled in this case.
-            // Clear the timeline and wait for the user to explicitly retry.
+
+        SendMessageResult::SendingFailed { send_error, txn_id } => {
+            timeline
+                .update_event_send_state(
+                    &txn_id,
+                    EventSendState::SendingFailed { error: Arc::new(send_error) },
+                )
+                .await;
+
+            // Clear the queue and wait for the user to explicitly retry (which will
+            // re-append the to-be-sent events in the queue).
             queue.clear();
         }
+
         SendMessageResult::TaskError { join_error, txn_id } => {
             error!("Message-sending task failed: {join_error}");
-            queue.clear();
 
-            let send_state = EventSendState::SendingFailed {
-                // FIXME: Probably not exactly right
-                error: Arc::new(matrix_sdk::Error::InconsistentState),
-            };
-            timeline_inner.update_event_send_state(&txn_id, send_state).await;
+            timeline
+                .update_event_send_state(
+                    &txn_id,
+                    EventSendState::SendingFailed {
+                        // FIXME: Probably not exactly right
+                        error: Arc::new(matrix_sdk::Error::InconsistentState),
+                    },
+                )
+                .await;
+
+            // See above comment in the `SendingFailed` arm.
+            queue.clear();
         }
     }
 }
 
 /// Result of [`SendMessageTask`].
 enum SendMessageResult {
-    /// The message was sent successfully, and the local echo was updated to
```

Comment:
Hmm, why is the additional context about the local echo not useful? Is it untrue or what?

Reply:
Indeed, it's not updated anymore. Now the local echo is updated in `handle_send_result` for all the match variants, which seemed clearer to me.